File: heron/stmgr/tests/cpp/grouping/custom-grouping_unittest.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <list>
#include <set>
#include <vector>
#include "gtest/gtest.h"
#include "proto/messages.h"
#include "basics/basics.h"
#include "errors/errors.h"
#include "threads/threads.h"
#include "network/network.h"

#include "basics/modinit.h"
#include "errors/modinit.h"
#include "threads/modinit.h"
#include "network/modinit.h"

#include "grouping/grouping.h"
#include "grouping/custom-grouping.h"

// Custom grouping is done at the instance level.
// The stmgr has no role here. The stmgr should not
// chose any tasks
TEST(CustomGrouping, test_nullop) {
  std::vector<sp_int32> task_ids;
  task_ids.push_back(0);
  task_ids.push_back(2);
  task_ids.push_back(4);
  task_ids.push_back(8);

  heron::stmgr::CustomGrouping* g = new heron::stmgr::CustomGrouping(task_ids);
  heron::proto::system::HeronDataTuple tuple;
  tuple.add_values("dummy");

  std::set<sp_int32> all_dests;
  for (sp_int32 i = 0; i < 1000; ++i) {
    std::vector<sp_int32> dests;
    g->GetListToSend(tuple, dests);
    EXPECT_EQ(dests.size(), (sp_uint32)0);
  }

  delete g;
}

int main(int argc, char** argv) {
  heron::common::Initialize(argv[0]);
  testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: heron/instance/tests/python/utils/custom_grouping_unittest.py
#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

# pylint: disable=missing-docstring

import unittest
from heron.instance.src.python.utils.misc import CustomGroupingHelper
from heron.instance.tests.python.utils.mock_generator import MockCustomGrouping

class CustomGroupingTest(unittest.TestCase):
  STREAM_ID = "default"
  TASK_IDS = [2, 3, 5, 7, 11, 13, 17]
  SRC_COMP_NAME = "component1"
  VALUES = ['test', 'values']

  def setUp(self):
    self.grouper = CustomGroupingHelper()

  def tearDown(self):
    self.grouper = None

  def test_all_target(self):
    # custom grouping returns all targets
    all_grouping = MockCustomGrouping(MockCustomGrouping.ALL_TARGET_MODE)
    self.grouper.add(self.STREAM_ID, self.TASK_IDS, all_grouping, self.SRC_COMP_NAME)
    self.assertIn(self.STREAM_ID, self.grouper.targets)

    self.grouper.prepare(None)
    ret = self.grouper.choose_tasks(self.STREAM_ID, self.VALUES)
    self.assertEqual(ret, self.TASK_IDS)
    ret2 = self.grouper.choose_tasks("not_registered_stream", self.VALUES)
    self.assertIsNotNone(ret2)
    self.assertIsInstance(ret2, list)
    self.assertEqual(len(ret2), 0)

  def test_random_target(self):
    # custom grouping returns a random subset of given targets
    random_grouping = MockCustomGrouping(MockCustomGrouping.RANDOM_TARGET_MODE)
    self.grouper.add(self.STREAM_ID, self.TASK_IDS, random_grouping, self.SRC_COMP_NAME)
    self.grouper.prepare(None)
    ret = self.grouper.choose_tasks(self.STREAM_ID, self.VALUES)
    self.assertEqual(len(set(ret)), len(ret))
    self.assertTrue(set(ret) <= set(self.TASK_IDS))

  def test_wrong_return_type(self):
    wrong_rettype = MockCustomGrouping(MockCustomGrouping.WRONG_RETURN_TYPE_MODE)
    self.grouper.add(self.STREAM_ID, self.TASK_IDS, wrong_rettype, self.SRC_COMP_NAME)
    self.grouper.prepare(None)

    with self.assertRaises(TypeError):
      self.grouper.choose_tasks(self.STREAM_ID, self.VALUES)

  def test_wrong_return_value(self):
    wrong_retvalue = MockCustomGrouping(MockCustomGrouping.WRONG_CHOOSE_TASK_MODE)
    self.grouper.add(self.STREAM_ID, self.TASK_IDS, wrong_retvalue, self.SRC_COMP_NAME)
    self.grouper.prepare(None)

    with self.assertRaises(ValueError):
      self.grouper.choose_tasks(self.STREAM_ID, self.VALUES)
